### ISSUE TITLE
asyncmy major backwards-incompatible change to TIME datatype in the 0.2.4 point release

### ISSUE THREAD

**Issue by zzzeek** [long2ice/asyncmy#37]
hey there -

it looks like asyncmy 0.2.4 suddenly changed the return format for TIME datatypes from a datetime.timedelta() , which is for whatever reason universally returned from all MySQL drivers, to now return the fully formed time() object, which is nice, but this is *extremely* backwards incompatible and belongs in asyncmy 0.3 at the very earliest, and preferably not at all by default as no other MySQL driver does this.   See example below; all drivers return a timedelta().   It would be very helpful if this behavior by default could be reverted and perhaps made some kind of option.

test suite for asyncmy:

```py

import asyncmy
import datetime


def time_from_mysql_delta(value):
    # convert from a timedelta value
    if value is not None:
        microseconds = value.microseconds
        seconds = value.seconds
        minutes = seconds // 60
        return datetime.time(
            minutes // 60,
            minutes % 60,
            seconds - minutes * 60,
            microsecond=microseconds,
        )
    else:
        return None

async def main():
    conn = await asyncmy.connect(user="scott", password="tiger", host="localhost", database="test")
    cursor = conn.cursor()

    await cursor.execute("""
create table if not exists time_test (data time)
""")

    data = datetime.time(12, 15, 23)
    await cursor.execute("delete from time_test")
    await cursor.execute("insert into time_test (data) values(%s)", (data, ))

    await cursor.execute("select data from time_test")
    row = await cursor.fetchone()
    assert time_from_mysql_delta(row[0]) == data

import asyncio
asyncio.run(main())
```

here's mysqlclient, pymysql, mariadbconnector, mysql connector; they all return a timedelta

```py

import datetime



def time_from_mysql_delta(value):
    # convert from a timedelta value
    if value is not None:
        microseconds = value.microseconds
        seconds = value.seconds
        minutes = seconds // 60
        return datetime.time(
            minutes // 60,
            minutes % 60,
            seconds - minutes * 60,
            microsecond=microseconds,
        )
    else:
        return None

def go(driver):
    conn = driver.connect(user="scott", password="tiger", host="localhost", database="test")
    cursor = conn.cursor()

    cursor.execute("""
drop table if exists time_test
""")
    cursor.execute("""
create table time_test (data time)
""")

    data = datetime.time(12, 15, 23)
    cursor.execute("delete from time_test")
    cursor.execute("insert into time_test (data) values(%s)", (data, ))

    print(f"Running driver: {driver}")
    cursor.execute("select data from time_test")
    row = cursor.fetchone()
    print(f"row received: {row}")
    time_from_mysql_delta(row[0]) == data

    conn.close()


import MySQLdb as mysqlclient
import pymysql
import mariadb
from mysql import connector as mysql


for driver in [mysqlclient, mariadb, pymysql, mysql]:
    go(driver)
```

output:

```
Running driver: <module 'MySQLdb' from '/home/classic/.venv3/lib/python3.10/site-packages/MySQLdb/__init__.py'>
row received: (datetime.timedelta(seconds=44123),)
Running driver: <module 'mariadb' from '/home/classic/.venv3/lib/python3.10/site-packages/mariadb/__init__.py'>
row received: (datetime.timedelta(seconds=44123),)
Running driver: <module 'pymysql' from '/home/classic/.venv3/lib/python3.10/site-packages/pymysql/__init__.py'>
row received: (datetime.timedelta(seconds=44123),)
Running driver: <module 'mysql.connector' from '/home/classic/.venv3/lib/python3.10/site-packages/mysql/connector/__init__.py'>
row received: (datetime.timedelta(seconds=44123),)
```

thanks!

**Comment by long2ice** [comment]
Why time type in db need return `datetime.timedelta`?

**Comment by zzzeek** [comment]
> Why time type in db need return `datetime.timedelta`? I'm not sure what kind of consideration is that.

I'm not either.   Whether there is some technical reason all drivers return a timedelta, or if it's because they are all striving to keep 100% compatibility with the original "MySQLdb" driver, is not fully clear, though it seems very likely that they are all just trying to be fully compatible with MySQLdb.      It seems very strange though that even mysql-connector and mariadb-connector, which are two drivers that are trying to do things their own way and are definitely not fully MySQLdb compatible anyway, are still doing this.

However, none of that is the "bug" here.  The bug is, [asyncmy has about 3000 downloads a week](https://pypistats.org/packages/asyncmy) right now.    So of the 3000 downloads and at least many hundreds of users, we assume some are querying MySQL TIME datatypes.   Their driver returned a `timedelta()`, which is weird, but OK, lets write all the code to interpret those `timedelta()` objects, as you see above in my sample where we did this in SQLAlchemy.

in everyone's requirements files, they're going to have a requirement for `asyncmy` probably without a version, since we are in 0.x numbers here.   They might have `asyncmy<0.3`.   Who knows.  But the main thing is, when you bump the version number from `x.y.z1` to `x.y.z2`, it is considered to be very poor form to **change** the API for something, such that code which calls upon that API will definitely break.     Any application that's out in the world which receives a `timedelta` from a TIME column, then they do a new install with ``pip install asyncmy``, will now break.    For us, it broke our CI, but for others, it could break in production, it could break for everyone downloading their software product, it's a big deal.

The main thing I'm trying to get across is that all changes have to be vetted for if they will break existing code or not.  and if so, you should not release a minor point release with that change, and ideally you would provide a migration path for the change so that the old behavior is deprecated, or emits a warning, something like that.

in this case, to change the return type here is extremely difficult.   I can't think of a good way for this to be a non-breaking change, except for a parameter someone would send to connect(), which in turn , nobody would really know about or ever use it.  That's probably why none of the drivers have changed it.    

so to sum up:  it's not about whether the new behavior is better or not.  this change is clearly better behavior.  however, it's backwards incompatible.

asyncmy is not that widely used yet and if you decide to keep it this way, I will go and change things on my end, as I will soon be hearing about this change anyway as a bug in my library.    it's not that big of a deal.   but as things get more popular, please be careful when making breaking changes, that's all!  thanks!

**Comment by long2ice** [comment]
OK, thanks for your replay, maybe rollback it in 0.2.5 is better

**Comment by long2ice** [comment]
I read the official docs in https://dev.mysql.com/doc/refman/8.0/en/time.html, now I know `datetime.timedelta` is right
